File: 1408-find-the-smallest-divisor-given-a-threshold/1408-find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:

bool calculateResult(vector<int> &nums, int devisor, int threshold)
{
    int result = 0;
    for (int i = 0; i < nums.size(); i++)
    {
        result += ceil(double(nums[i]) / double(devisor));
    }
    return result <= threshold;
}

int smallestDivisor(vector<int> &nums, int threshold)
{
    if (nums.size() > threshold)
        return -1;
    int high = INT_MIN;
    for (int i = 0; i < nums.size(); i++)
        high = max(high, nums[i]);
    int low = 1;
    while (low <= high)
    {
        int mid = low + (high - low) / 2;
        if (calculateResult(nums, mid, threshold))
        {
            high = mid - 1;
        }
        else
            low = mid + 1;
    }
    return low;
}
};

File: 1408-find-the-smallest-divisor-given-a-threshold/1408-find-the-smallest-divisor-given-a-threshold.java
class Solution {

    Boolean calculate(int[] nums, int divisor, int threshold) {
        int ans = 0;
        for (int i = 0; i < nums.length; i++) {
            ans += (int) Math.ceil((double) nums[i] / (double) divisor);
        }
        return ans <= threshold;
    }

    public int smallestDivisor(int[] nums, int threshold) {
        int low = 1;
        int high = Arrays.stream(nums).max().getAsInt();
        int ans = high;
        while (low <= high) {
            int mid = (low + high)>>1;
            if (calculate(nums, mid, threshold)) {
                ans = mid; 
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }
}